query_evaluation.py

```diff
@@ -78,52 +80,106 @@ def lm_evaluate_query(query: List[str]):
+        ranks = [utility.rankify(dict(enumerate(x))) for x in matrices[i]]
+        if i < 4:
+            for query_n, (answer, _) in enumerate(queries[i]):
+                # GTP is answer
+                if ranks[query_n].index(answer) < X:
+                    precision.append(1/X)
+        else:
+            with Pool(processes=8) as p:
+                max_ = len(list(enumerate(queries[i])))
+                with tqdm(total=max_) as pbar:
+                    for _, score in enumerate(p.starmap(partial(precision_function, ranks, X), list(enumerate(queries[i])))):
+                        precision.append(score)
+                        pbar.update()
+        precisions.append(np.mean(precision))
+        print(np.mean(precision))
+    return precisions
+
+
+def mean_average_precision(matrices):
+    """
+    Calculates the documents queries for the first 4 iterations and
+    then it calculates the topic queries for the last 4 iterations
+    :param matrices: the 8 query sets
+    :return: mean average precision
+    """
+    MAP = []
+    for i in range(8):  # 8 because there are 8 sets of queries of different lengths
         AP = []
         ranks = [utility.rankify(dict(enumerate(x))) for x in matrices[i]]
         if i < 4:
             for query_n, (answer, _) in enumerate(queries[i]):
                 # GTP is answer
                 AP.append(1 / (ranks[query_n].index(answer) + 1))
         else:
-            for query_n, (answer, _) in tqdm(list(enumerate(queries[i]))):
-                topic = dt_matrix.getcol(answer).toarray()
-                threshold = topic.mean()
-                gtp_ids = np.nonzero(np.where(topic < threshold, 0, topic))[0]
-                gtp_ranks = [ranks[query_n].index(gtp_id) for gtp_id in gtp_ids]
-                precision = [(i + 1) / (rank + 1) for i, rank in enumerate(gtp_ranks)]
-                AP.append(np.mean(precision))
+            with Pool(processes=8) as p:
+                max_ = len(list(enumerate(queries[i])))
+                with tqdm(total=max_) as pbar:
+                    for _, score in enumerate(p.starmap(partial(map_function, ranks), list(enumerate(queries[i])))):
+                        AP.append(score)
+                        pbar.update()
         MAP.append(np.mean(AP))
         print(np.mean(AP))
     return MAP
 
 
+def map_function(ranks, query_n, answer):
+    topic = dt_matrix.getcol(answer[0]).toarray()
+    threshold = topic.mean()
+    gtp_ids = np.nonzero(np.where(topic < threshold, 0, topic))[0]
+    gtp_ranks = [ranks[query_n].index(gtp_id) for gtp_id in gtp_ids]
+    gtp_ranks.sort()
+    precision = [(i + 1) / (gtp_ranks[i] + 1) for i in range(len(gtp_ranks))]
+    return np.mean(precision)
+
+
+def precision_function(ranks, X, query_n, answer):
+    topic = dt_matrix.getcol(answer[0]).toarray()
+    threshold = topic.mean()
+    gtp_ids = np.nonzero(np.where(topic < threshold, 0, topic))[0]
+    gtp_in_N = [x for x in ranks[query_n][:X] if x in gtp_ids]
```

Comment:
Ved stadig ikke hvad gtp står for, da det aldrig beskrives. Kommentar eller lignende.